### PR TITLE
Fix logits mem leak

### PR DESCRIPTION
**Pull request by garrett361**
The current code leaves `logits`, `shift_logits`, and `outputs` all defined throughout every loop.  This means that references to the logits from the previous iteration are still alive while progressing through the next iteration, and so their CUDA memory cannot be freed and re-used. Logits are very memory intensive: for a 128k vocab size and 4k seqlen, `bfloat16` logits cost `1 GiB * batch_size`, so this should be avoided.

This PR deletes these references as soon as possible, thereby freeing the memory.

### CODE REVIEW

**Comment by fabianlim** [comment]
this should only affect sum loss right

**Comment by garrett361** [comment]
Should be both because even in the `mean` case, `outputs` is held onto.

**Review by fabianlim — APPROVED** [review]
@garrett361 thanks for the test. It does show that there is an advantage in some instances. Approving.

**Comment by garrett361** [comment]
The plots settled into something a little more informative looking after running for a few hours. Same coloring as before, but with a curve for every gpu on the same plot:
Scherm­afbeelding 2025-07-11 om 8 16 16 PM
A clear trend of less memory usage per GPU on average after this commit.

FYI @fabianlim @dangxuanhong
